File: MemoryGame/Engine/MatchStrategy.swift
//
//  MatchStrategy.swift
//  Memory Match Kids
//

import Foundation

protocol MatchStrategy {
    func isMatch(cards: [CardModel], mode: MatchMode) -> Bool
    func requiredSelectionCount(mode: MatchMode) -> Int
}

struct NormalMatchStrategy: MatchStrategy {
    func requiredSelectionCount(mode: MatchMode) -> Int { 2 }

    func isMatch(cards: [CardModel], mode: MatchMode) -> Bool {
        guard cards.count == 2 else { return false }
        let groupA = cards[0].groupId ?? cards[0].pairId
        let groupB = cards[1].groupId ?? cards[1].pairId
        if groupA == groupB { return true }
        if cards[0].content.id == cards[1].content.id { return true }
        return false
    }
}

struct AssociationMatchStrategy: MatchStrategy {
    func requiredSelectionCount(mode: MatchMode) -> Int { 2 }

    func isMatch(cards: [CardModel], mode: MatchMode) -> Bool {
        guard cards.count == 2 else { return false }
        let first = cards[0].content
        let second = cards[1].content
        guard first.id != second.id else { return false }

        let groupA = cards[0].groupId ?? cards[0].matchingPairId ?? cards[0].pairId
        let groupB = cards[1].groupId ?? cards[1].matchingPairId ?? cards[1].pairId
        if groupA == groupB { return true }

        return MathAssociationHelper.expressionMatchesAnswer(first.label, second.label)
    }
}

struct TripleMatchStrategy: MatchStrategy {
    func requiredSelectionCount(mode: MatchMode) -> Int { 3 }

    func isMatch(cards: [CardModel], mode: MatchMode) -> Bool {
        guard cards.count == 3 else { return false }
        let gid = cards[0].groupId ?? cards[0].matchingPairId ?? cards[0].pairId
        return cards.allSatisfy {
            ($0.groupId ?? $0.matchingPairId ?? $0.pairId) == gid
        }
    }
}

enum MatchStrategyFactory {
    static func strategy(for mode: MatchMode) -> MatchStrategy {
        switch mode {
        case .identical, .sequence, .maze:
            return NormalMatchStrategy()
        case .association:
            return AssociationMatchStrategy()
        case .triple:
            return TripleMatchStrategy()
        }
    }
}

File: android/app/src/main/java/com/memogame/app/engine/MatchStrategy.kt
package com.memogame.app.engine

import com.memogame.app.model.CardModel
import com.memogame.app.model.MatchMode

interface MatchStrategy {
    fun isMatch(cards: List<CardModel>): Boolean
    val requiredSelectionCount: Int
}

class NormalMatchStrategy : MatchStrategy {
    override val requiredSelectionCount = 2

    override fun isMatch(cards: List<CardModel>): Boolean {
        if (cards.size != 2) return false
        val groupA = cards[0].groupId ?: cards[0].pairId
        val groupB = cards[1].groupId ?: cards[1].pairId
        if (groupA == groupB) return true
        return cards[0].content.id == cards[1].content.id
    }
}

class AssociationMatchStrategy : MatchStrategy {
    override val requiredSelectionCount = 2

    override fun isMatch(cards: List<CardModel>): Boolean {
        if (cards.size != 2) return false
        val first = cards[0].content
        val second = cards[1].content
        if (first.id == second.id) return false

        val groupA = cards[0].groupId ?: cards[0].matchingPairId
        val groupB = cards[1].groupId ?: cards[1].matchingPairId
        if (groupA == groupB) return true

        return MathAssociationHelper.expressionMatchesAnswer(first.label, second.label)
    }
}

class TripleMatchStrategy : MatchStrategy {
    override val requiredSelectionCount = 3

    override fun isMatch(cards: List<CardModel>): Boolean {
        if (cards.size != 3) return false
        val gid = cards[0].groupId ?: cards[0].matchingPairId
        return cards.all { (it.groupId ?: it.matchingPairId) == gid }
    }
}

object MatchStrategyFactory {
    fun strategy(mode: MatchMode): MatchStrategy = when (mode) {
        MatchMode.IDENTICAL -> NormalMatchStrategy()
        MatchMode.ASSOCIATION -> AssociationMatchStrategy()
        MatchMode.TRIPLE -> TripleMatchStrategy()
    }
}
